packages/datadog-instrumentations/src/cassandra-driver.js

```diff
@@ -0,0 +1,207 @@
+      if (!startCh.hasSubscribers) {
+        return _innerExecute.apply(this, arguments)
+      }
+      const asyncResource = new AsyncResource('bound-anonymous-fn')
+      const isValid = (args) => {
+        return args.length === 4 || typeof args[3] === 'function'
+      }
+
+      if (!isValid(arguments)) {
+        return _innerExecute.apply(this, arguments)
+      }
+
+      startCh.publish({ client: this, query })
+
+      const lastIndex = arguments.length - 1
+      let cb = arguments[lastIndex]
+
+      if (typeof cb === 'function') {
+        cb = asyncResource.bind(cb)
+        arguments[lastIndex] = AsyncResource.bind(wrapCallback(asyncEndCh, errorCh, cb))
+      }
+
+      try {
+        return _innerExecute.apply(this, arguments)
+      } catch (e) {
+        finish(asyncEndCh, errorCh, e)
+        throw e
+      } finally {
+        endCh.publish(undefined)
+      }
+    }
+  )
+  return cassandra
+})
+
+addHook({ name: 'cassandra-driver', versions: ['>=3.3'], file: 'lib/request-execution.js' }, RequestExecution => {
+  const addHostCh = channel(`apm:cassandra:general:addHost`)
+  shimmer.wrap(RequestExecution.prototype, '_sendOnConnection', _sendOnConnection => function () {
+    addHostCh.publish({ connectionOptions: this._connection })
+    return _sendOnConnection.apply(this, arguments)
+  })
+  return RequestExecution
+})
+
+addHook({ name: 'cassandra-driver', versions: ['3.3 - 4.3'], file: 'lib/request-execution.js' }, RequestExecution => {
+  const addHostCh = channel(`apm:cassandra:general:addHost`)
+  shimmer.wrap(RequestExecution.prototype, 'start', start => function (getHostCallback) {
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    const execution = this
+
+    if (!isRequestValid(this, arguments, 1)) {
+      return start.apply(this, arguments)
+    }
+
+    getHostCallback = asyncResource.bind(getHostCallback)
+
+    arguments[0] = AsyncResource.bind(function () {
+      addHostCh.publish({ connectionOptions: execution._connection })
+      return getHostCallback.apply(this, arguments)
+    })
+
+    return start.apply(this, arguments)
+  })
+  return RequestExecution
+})
+
+addHook({ name: 'cassandra-driver', versions: ['3 - 3.2'], file: 'lib/request-handler.js' }, RequestHandler => {
+  const addHostCh = channel(`apm:cassandra:general:addHost`)
+  shimmer.wrap(RequestHandler.prototype, 'send', send => function (request, options, callback) {
+    const handler = this
+
+    if (!isRequestValid(this, arguments, 3)) {
+      return send.apply(this, arguments)
+    }
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+
+    callback = asyncResource.bind(callback)
+
+    arguments[2] = AsyncResource.bind(function () {
+      addHostCh.publish({ connectionOptions: handler.connection })
+      return callback.apply(this, arguments)
+    })
+
+    return send.apply(this, arguments)
+  })
+  return RequestHandler
+})
+
+function combine (queries) {
```

Comment:
As mentioned above, this should only be here if the result of this is what is actually sent to Cassandra.